### ISSUE TITLE
Configuration drift not detected for changes to Proof Key for Code Exchange Code Challenge Method (PKCE)

### ISSUE THREAD

**Issue by sharpedavid** [keycloak/terraform-provider-keycloak#700]
I think this is arguably a feature request because I assume configuration drift is detected on a best-effort basis.

If I manually change the [Proof Key for Code Exchange Code Challenge Method](https://registry.terraform.io/providers/mrparkers/keycloak/latest/docs/resources/openid_client#pkce_code_challenge_method) of a `keycloak_openid_client` Resource using the Keycloak Admin Console, running `terraform plan` does not detect the changes.

`terraform plan` output:
![image](https://user-images.githubusercontent.com/1767127/175423251-d8725079-0480-4104-a821-09a09cc75882.png)

Keycloak Admin Console:
![image](https://user-images.githubusercontent.com/1767127/175423040-f72f0d3a-9198-4d4a-9826-105dc44e5677.png)

Compare this to if I manually change the client "Description", in which case `terraform plan` will detect and report the difference:

![image](https://user-images.githubusercontent.com/1767127/175422890-4fc6cd36-d610-4d99-97a0-c1f6dad28c91.png)

My expectation is that Terraform would detect the manual change made to the `pkce_code_challenge_method` attribute and plan to override it.

**Comment by dmeyerholt** [comment]
Hi @sharpedavid ,
currently PKCE parameters are not "first-class-citizens" in terms of dedicated parameters to use in the `keycloak_oidc_identity_provider` resource. You can however detect configuration drift and/or configure pkce using the `extra_config` block. eg:
```
resource "keycloak_oidc_identity_provider" "idp" {
.....
extra_config = {
    "clientAuthMethod" = "client_secret_post"
    "pkceEnabled" = true
    "pkceMethod" = "S256"
  }
....
}
```

This yields for example to drift detection like this:
```
  ~ resource "keycloak_oidc_identity_provider" "idp" {
      ~ extra_config                            = {
          ~ "pkceEnabled"      = "false" -> "true"
          ~ "pkceMethod"       = "plain" -> "S256"
            # (1 unchanged element hidden)
        }
        id                                      = "raiffeisen"
        # (25 unchanged attributes hidden)
    }

```

**Comment by alexmeise** [comment]
Is it normal to still need client_secret even when PKCE is enabled? 


```
extra_config = {
  "display_name" = "Login with SSO"
  "pkceEnabled" = true
  "pkceMethod" = "S256"
}
```

If I dont pass a client_secret i get: 

```
╷
│ Error: Missing required argument
│
│   on ../../../_modules/master/main.tf line 49, in resource "keycloak_oidc_identity_provider" "master_realm_identity_provider":
│   49: resource "keycloak_oidc_identity_provider" "master_realm_identity_provider" {
│
│ The argument "client_secret" is required, but no definition was found.
╵
```

One of the reasons to use PKCE is to not have to use a secret at all.

Is there any way around this?

**Comment by sschu** [comment]
This is not the purpose, to quote https://oauth.net/2/pkce/: "PKCE is not a form of client authentication, and PKCE is not a replacement for a client secret or other client authentication. PKCE is recommended even if a client is using a client secret or other form of [client authentication](https://oauth.net/2/client-authentication/) like [private_key_jwt](https://oauth.net/private-key-jwt)."

**Comment by alexmeise** [comment]
> This is not the purpose, to quote https://oauth.net/2/pkce/: "PKCE is not a form of client authentication, and PKCE is not a replacement for a client secret or other client authentication. PKCE is recommended even if a client is using a client secret or other form of [client authentication](https://oauth.net/2/client-authentication/) like [private_key_jwt](https://oauth.net/private-key-jwt)."

Thanks for the information. I understand.

**Comment by alexmeise** [comment]
Just a note, having private_key_jwt and PKCE enabled does indeed allow client authentication without secret. Even if it is not the purpose.

**Comment by sschu** [comment]
Yup. But I assume Keycloak currently does not support this for integration with external IDPs via OIDC.

**Comment by sharpedavid** [comment]
> Just a note, having private_key_jwt and PKCE enabled does indeed allow client authentication without secret. Even if it is not the purpose.

I think you're mixing things up, @alexmeise. You originally asked about PKCE, but you didn’t mention private_key_jwt. private_key_jwt is an implementation of [RFC 7523](https://www.rfc-editor.org/rfc/rfc7523.txt) and does not use a secret—that’s the whole point. @sschu was correct in saying that PKCE is not a form of client authentication, but private_key_jwt is.
